### ISSUE TITLE
Driver crashing with specific situations

### ISSUE THREAD

**Issue by PenguinKeeper7** [hashcat/hashcat#3897]
**Describe the bug**
Drivers crashing in very specific scenarios. Hashcat says "Initializing backend runtime for device #1. Please be patient...", then exits with no other text or error. This seems to happen very rarely but when it does happen, it is consistent and this bug has been experienced by multiple people. The only visible error is in Windows' Event Viewer: 
```
Faulting application name: hashcat.exe, version: 0.0.0.0, time stamp: 0x6539dbcf
Faulting module name: nvcuda64.dll, version: 31.0.15.4584, time stamp: 0x65270942
```

There appears to be 3 possible behaviours:
1) Works normally, as expected (This is the behaviour for 99% of attacks)
2) Tick-tocking between crashing and passing the "Initializing backend ..." message. The same exact command can fail, then succeed, then fail, then succeed etc. In both situations, the kernel is built in the `kernels/` folder but seemingly fails to be loaded onto the compute device and causes the driver to crash.
3) Consistent failure, no matter how many times the same command is run, it always crashes on "Initializing backend ...".

It appears that both the number of rules and the number of hashes have some play this. In my case, it was consistently failing above 16 rules and tick-tocked from 0-16 rules. and the same with the amount of hashes, 8448 or above SHA256 hashes would lead to consistent failure, while 1-8447 would tick-tock. Unknown if this number is consistent across all failing situations, as said previously, it's very difficult to get this state so getting test cases is tough.

This problem seems to be something strange with compilation, building with cygwin's `gcc` makefile seems to be fine, while building with WSL's `x86_64-w64-mingw32-gcc` experiences this behaviour, likely why the majority of people haven't experienced this and is only limited to people building without Cygwin, identical to https://github.com/hashcat/hashcat/issues/3870.

It seems to have started with this commit: https://github.com/hashcat/hashcat/commit/ab932815eaa94b5ff1c64320ba0387c096675027 however, that code change seems innocent so the problem is likely elsewhere, where those variables are being used, however, the new code does consistently create significantly higher `size_rules`. Reverting this change leads to consistent success. Release 6.2.6 does not show this behaviour, while GitHub master does.

**To Reproduce**
Compile preferably without Cygwin. I am using `make win`, `make` is untested and may not exhibit the same behaviour. I'm happy to provide pre-compiled binaries if needed due to this being sporadic with environments/compilation.
```./hashcat -m 1400 xhashes.txt empty.txt -r test.rule```
[testFiles.zip](https://github.com/hashcat/hashcat/files/13191840/testFiles.zip)

**Hardware/Compute device:**
```
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 545.84                 Driver Version: 545.84       CUDA Version: 12.3     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                     TCC/WDDM  | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA GeForce RTX 3070      WDDM  | 00000000:08:00.0  On |                  N/A |
|  0%   48C    P8              17W / 220W |    638MiB /  8192MiB |      2%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
```

**Relevant Discord conversations**:
https://discord.com/channels/566331281558732800/874753308953280613/1150887195784269834
https://discord.com/channels/566331281558732800/874753308953280613/1152025811755487292
https://discord.com/channels/566331281558732800/874753354511843368/1166910678687305748

**Compiler versions**:
Cygwin (Passes) - `gcc (GCC) 11.3.0`
WSL (Fails) - `x86_64-w64-mingw32-gcc (GCC) 10-win32 20220113` running `Ubuntu 22.04.2 LTS`

**Comment by PenguinKeeper7** [comment]
Minor update - It seems to happen disproportionally more often on phpass (-m 400) for some reason - unsure why. It does happen on other modes too but significantly more often on phpass

**Comment by matrix** [comment]
@PenguinKeeper7 could you try applying #4115 to see if it also solves the problem you found?

**Comment by PenguinKeeper7** [comment]
@matrix, yeah Chicken found this a while ago but while it does drastically help, it certainly doesn't fix the problem and very sporadic crashes still occur, so there may be yet more out-of-bounds reading. Recreation steps are pretty much impossible

**Comment by matrix** [comment]
@jsteube I think #4115 should be merged as soon as possible.
@PenguinKeeper7 have you tried with DEBUG=2 ? If hashcat crashes a lot of useful info is printed to identify the reason

**Comment by PenguinKeeper7** [comment]
Hashcat doesn't crash in the typical sense, the OS force-kills it with 0 loggability. The only thing you can ever get out of it is an address reading violation. I've seen 5-10 others have this issue, although I seem to have by far the most number of these crashes. I usually got upwards of 10 per day without a similar patch (not this exact one, but same idea) and I usually get 0-1 per day now. Debugging is near impossible, profilers show nothing, other than that same access violation

Info from January 2025 that I posted in the discord:
```
24th - 8
25th - 8
26th - 8
27th - 13
28th - 10
29th - 7
---- Patch applied ----
30th - 1
Today - 0
```
https://discord.com/channels/566331281558732800/874753354511843368/1334926970106151003

**Comment by Chick3nman** [comment]
@PenguinKeeper7 can you retest from current and see if you can cause this crash? #4115 does appear to address the buffer problem we had identified previously

**Comment by PenguinKeeper7** [comment]
@Chick3nman I've mostly used post-patch binaries (both our janky patch and #4115's admittedly better one) for the last few months and I haven't noticed almost any driver crashing with identical behaviour to this one. I still have experienced a very, very small number of rare crashes that feel similar, so there may be another out-of-bounds read somewhere or simply just an unrelated environmental problem but just re-running Hashcat works so replicating it is basically impossible. I'll close this because I think this definitely is solved and the many people that have had this issue have also reported that our patch or #4115 fixed it.
